### PR TITLE
Feature/fetch past deals pagination

### LINKED ISSUE #58: Enable pagination for deals
Allow fetching of past deals with pagination.

Use query params (from, to) for pages.

### PR DESCRIPTION
**Pull request by nmlinaric**
Created endpoint ```/user/node/pastdeals/:nodeId``` for fetching node past deals.
Endpoint has pagination (from=recordNumber&to=recordNumber) and response data ordering (ascending or descending).

Resolves #58 Enable pagination for deals.

### CODE REVIEW

File: src/Controller/Api/NodePastDealsControllerValidation.ts [hunk 1]
```diff
@@ -33,3 +32,17 @@ export const CreateNodePastDealsValidationSchema = Joi.object({
...
+        from: number;
+        to: number;
+        orderBy: string;
+    };
+}
+
+export const FetchNodePastDealsValidationSchema = Joi.object({
+    from: Joi.number().required(),
+    to: Joi.number().required(),
+    orderBy: Joi.string().uppercase().valid('ASC', 'DESC').required()
```
Comment (morrigan) [inline src/Controller/Api/NodePastDealsControllerValidation.ts]: We don't need to be so strict here. Put ASC by default and all good

File: src/Controller/Api/NodePastDealsController.ts [hunk 2]
```diff
@@ -28,4 +28,20 @@ export class NodePastDealsController {
             res.status(500).json({error: "An unknown error occurred."});
         }
     }
+
+    public async fetchNodePastDeals(req: ValidatedRequest<FetchNodePastDealsRequestSchema>, res: Response) {
+        try {
+            const userId = res.locals.userId;
+            const result = await this.nodePastDealsService.fetchNodePastDeals(
```
Comment (morrigan) [inline src/Controller/Api/NodePastDealsController.ts]: you could use nice ES6 syntax `const {from, to, orderBy} = req.query`

File: src/Services/NodePastDealsService.ts [hunk 3]
```diff
@@ -28,4 +29,22 @@ export class NodePastDealsService {
         }
         return await NodePastDealModel.create({cid, state, size, provider, price, duration, nodeId});
     }
+
+    public async fetchNodePastDeals(userId: number, from: number, to: number, orderBy: string) {
+        return await NodePastDealModel.findAll({
+            attributes: ['id', 'cid', 'state', 'size', 'provider',
```
Comment (morrigan) [inline src/Services/NodePastDealsService.ts]: any reason why not all attributes but specifically hardcoded here?

File: src/Services/NodePastDealsService.ts [hunk 4]
```diff
@@ -28,4 +29,22 @@ export class NodePastDealsService {
...
+                'price', 'duration', 'createdAt', 'updatedAt'],
+            include: [{
```
Comment (morrigan) [inline src/Services/NodePastDealsService.ts]: No need for the include, we don't use it in other endpoints either so this shouldn't be a special case

File: src/Controller/Api/NodePastDealsControllerValidation.ts [hunk 5]
```diff
@@ -33,3 +32,17 @@ export const CreateNodePastDealsValidationSchema = Joi.object({
...
+export interface FetchNodePastDealsRequestSchema extends ValidatedRequestSchema {
+    [ContainerTypes.Query]: {
+        from: number;
+        to: number;
+        orderBy: string;
+    };
+}
+
+export const FetchNodePastDealsValidationSchema = Joi.object({
+    from: Joi.number().required(),
```
Comment (morrigan) [inline src/Controller/Api/NodePastDealsControllerValidation.ts]: `Joi.number().min(0)` so we don't include negative numbers

File: src/Controller/Api/NodePastDealsControllerValidation.ts [hunk 6]
```diff
@@ -33,3 +32,17 @@ export const CreateNodePastDealsValidationSchema = Joi.object({
...
+    to: Joi.number().required(),
```
Comment (morrigan) [inline src/Controller/Api/NodePastDealsControllerValidation.ts]: should be bigger than `from`

File: src/Services/NodePastDealsService.ts (hunk 3 above)
Comment (nmlinaric) [reply src/Services/NodePastDealsService.ts]: At some point I had to do this to retrieve only records from this table without the data from nodes table. Now with the include below this is no longer needed, my oversight :)

File: src/Services/NodePastDealsService.ts (hunk 4 above)
Comment (nmlinaric) [reply src/Services/NodePastDealsService.ts]: The idea was to retrieve only past deals which belong to a node that belongs to the currently logged in user. Little bit confusing syntax, but the only way to achieve that.

**Pull request merged**